text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,172 @@
+
+Each time the local repository has been updated via `git pull`, buildapack data will be processed/normalized into a single JSON object, where plain text searches can be used against it.   Fields in this JSON object will be re-indexed as searchable fields, which will be compared against the search text during the retrieval algorithm.
+
+In addition, a server process will expose endpoints and handle incoming request for the GET endpoints mentioned earlier.  For search requests, the text will be extracted from the `query` query parameter and used to search against the normalized buildpacks index.   Results will be added to the server response, as a list of JSON objects.
+
+*Note:  The Distribution Team maintainers can manage this service, and could even set-up a pager (but only best effort, and during working hours).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- more vulnerable due to Github outages
+- less optimal than using actual DB technology
+- constraint by what's in the index registry repository.  We might want to augment the metadata, with new fields that don't makes sense to have in the index repository.  I think this maybe an issue for the `/buildpacks/1` endpoint that *could* have download data, etc. things that don't belong in the actual repo.
+- without a DB we can't easily do things like ORDER_BY, or use ASC|DESC
+- someone has to maintain the service
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+  - Do nothing
+  
+    We already have buildpack metadata stored in the [registry](https://github.com/buildpacks/registry-index).  
+    Customers can simply explore this repository for buildpack information.  Perhaps someone else in the community will
+    come up with their own solution that can be donated to CNB.
+
+  - Having a separate, independent service responsible for:
+    1. pulling from the registry-index repository
+    2. augmenting buildpack data
+    3. persisting it to a hosted PostgresDB  
+
+- Why is this proposal the best?
+
+  Seems like a good first iterative step that's easy to implement with less moving parts. 
+  Helps us to quickly get something out there for customers to experiment with (MVP).
+
+- What is the impact of not doing this?
+
+  Customers will struggle to find relevant buildpacks to aid them in their development.  
+ 
+ 
+# Prior Art
+[prior-art]: #prior-art
+
+  1. [Ruby Gems](https://rubygems.org/)
+  2. [npm](https://www.npmjs.com/)
+  3. [Rust](https://crates.io/)
+  4. [Dockerhub](https://hub.docker.com/)
+
+# Unresolved Questions
```

Comment:
Is this API part of the extension spec and if so should we look into versioning? We probably should do versions at least through headers but it's pretty common to do it via URL as well. Stripe has written a good [post on API versioning](https://stripe.com/blog/api-versioning) using dates instead of version numbers.

On top of that should we be looking into standards like OpenAPI or JSONAPI? There are some good tools like [swagger](https://swagger.io/) that can help here.

Reply:
Given very small number of endpoints on this server, I'm not terribly worried about selecting a standard such as OpenAPI or JsonAPI, but I'll take a look at these.   I need to think about the versioning.

Reply:
Re: versioning

My suggestion: just slap a v1 prefix to the base url. ie. `.../v1/buildpacks/....`

This also typically makes organization of the implementation easier.

Reply:
yep, that's the route I went down

Reply:
@jromero I added a subsection that calls this out.